// balancertags/src/getBalancerv2PoolTags.ts
// Use dynamic import for node-fetch
import fetch from "node-fetch";

// Updated Pool interface to match the new query structure
interface Pool {
  address: string; // Changed from 'id' to 'address'
  createTime: number;
  tokens: {
    symbol: string;
    name: string;
  }[];
}

interface Tag {
  "Contract Address": string;
  "Public Name Tag": string;
  "Project Name": string;
  "UI/Website Link": string;
  "Public Note": string;
}

// Updated to reflect the correct response structure based on the query
interface GraphQLResponse {
  data: {
    pools: Pool[];
  };
}

const GET_POOLS_QUERY = `
  query GetPools($lastTimestamp: Int) {
    pools(
      first: 1000,
      orderBy: createTime,
      orderDirection: asc,
      where: { createTime_gt: $lastTimestamp }
    ) {
      address
      createTime
      tokens {
        symbol
        name  
      }
    }
  }
`;

async function returnTags(
  chainId: number,
  apiKey: string
): Promise<Tag[] | Error> {
  let lastTimestamp: number = 0;
  let allTags: Tag[] = [];
  let isMore = true;

  if (chainId !== 1) {
    throw new Error(
      `Unsupported Chain ID: ${chainId}. This function only supports Ethereum Mainnet.`
    );
  }

  if (!apiKey || apiKey.length < 20) {
    throw new Error("Invalid API key format.");
  }

  while (isMore) {
    const response = await fetch(
      "https://api.thegraph.com/subgraphs/name/balancer-labs/balancer-v2",
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json",
          Authorization: `Bearer ${apiKey}`,
        },
        body: JSON.stringify({
          query: GET_POOLS_QUERY,
          variables: { lastTimestamp: lastTimestamp },
        }),
      }
    );

    if (!response.ok) {
      throw new Error(`Network response was not ok: ${response.statusText}`);
    }
    const result = (await response.json()) as GraphQLResponse;
    console.log(result.data);
    const pools: Pool[] = result.data.pools;

    allTags.push(...transformPoolsToTags(pools));

    if (pools.length < 1000) {
      isMore = false;
    } else {
      // Assuming 'createTime' is now a property on the returned pools, which needs verification
      lastTimestamp = parseInt(pools[pools.length - 1].createTime.toString());
    }
  }

  return allTags;
}

// Adjusted transformation function to match the updated Pool interface
function transformPoolsToTags(pools: Pool[]): Tag[] {
  return pools.map((pool) => ({
    "Contract Address": pool.address,
    "Public Name Tag": `${pool.tokens
      .map((t) => t.symbol)
      .join("/")} Pool`,
    "Project Name": "Balancer v2",
    "UI/Website Link": "https://balancer.fi/",
    "Public Note": `A Balancer v2 pool with tokens ${pool.tokens
      .map((t) => t.name)
      .join(", ")}.`,
  }));
}

export { returnTags };
